sample/peer/cmd/request.go

```diff
@@ -90,7 +92,9 @@ func request(req []byte) ([]byte, error) {
 
 	rc := connector.New()
 
-	err = rc.ConnectManyReplicas(peerAddrs, grpc.WithInsecure(), grpc.WithBlock())
+	ctx := context.Background()
+	ctx, _ = context.WithTimeout(ctx, 5*time.Second)
+	err = rc.ConnectManyReplicas(ctx, peerAddrs, grpc.WithInsecure(), grpc.WithBlock())
```

Comment:
I haven't tried, but I think `grpc.WithBackoffMaxDelay()` may be even omitted for simplicity. I guess gRPC simply does infinite back off by default.

Reply:
> Have you considered removing `grpc.WithBlock()` option? I tried to play a bit and replaced `grpc.WithBlock()` with `grpc.WithBackoffMaxDelay()`. Then added `grpc.FailFast(false)` to invocation of `h.client.Chat()` in `messageStreamHandler.HandleMessageStream()` of `sample/net/grpc/connector`, and wrapped most of its body in another goroutine. It seems to work... Maybe you could try experimenting a bit in this direction further?

I didn't think much about `grpc.WithBlock()`, so OK, I'll try to do in your approach next.

Reply:
My experiment with removing `grpc.WithBlock()` and using `grpc.FailFast(false)` has shown that it seems to be possible to effectively mask connectivity failures solely by means of gRPC framework.

More importantly, this is also a matter of _our external API_. As MinBFT paper mentions that network faults are always masked, now I tend to think that we might need to refine the description of `ReplicaConnector` and `MessageStreamHandler` interfaces in `api/api.go`. Namely, require `ReplicaConnector.ReplicaMessageStreamHandler` to always succeed and return immediately, if the supplied parameter `replicaID` is correct. Additionally, remove error return value from `MessageStreamHandler.HandleMessageStream`, so that it is not allowed to fail.

Informally, those changes in API would mean that it should always be possible to immediately construct a local _representation_ of a remote replica. Moreover such a representation would be guaranteed to always be able to accept a channel of incoming messages and return a channel of messages produced in reply. In case of a network fault, the representation would simply stop sending/receiving messages to/from those channels, and then continue again, should the failure be resolved. Thus, the core would be free from handling of any kind of connectivity failure.

Reply:
@sergefdrv hopefully the new version of PR is close to what you intended. I learned/tested some grpc options and finally think that I double-checked the suggested approach is quite better than v1. Thank you very much :smile: